src/Chainweb/Cut/Create.hs

```diff
@@ -0,0 +1,442 @@
+    _chainwebVersion = _chainwebVersion . _cutExtensionCut
+    {-# INLINE _chainwebVersion #-}
+
+-- | Witness that a cut can be extended for the given chain by trying to
+-- assemble the adjacent hashes for a new work header.
+--
+-- If there is a graph change the adajacent parents on the new chains are the
+-- genesis parents.
+--
+-- Graph changes add some complexity to this function. There are three cases:
+--
+-- 1. This is a transition cut, i.e. the cut contains block headers with
+--    different graphs. In this case we wait until all chains transitions to the
+--    new graph before add the genesis blocks for the new chains and move ahead.
+--
+-- 2. This is a transition step (parentGraph /= targetGraph). Only do the step
+--    when new /and/ old adjacents headers are available. The old dependencies
+--    are not used but are needed to guarantee that the cut can extended on
+--    other chains.
+--
+-- 3. This is a "normal" transition.
+--
+-- TODO: it is important that the semantics of this function corresponds to the
+-- respective validation in the module "Chainweb.Cut", in particular
+-- 'isMonotonicCutExtension'. It must not happen, that a cut passes validation
+-- that can't be further extended.
+--
+getCutExtension
+    :: HasCallStack
+    => HasChainId cid
+    => Cut
+        -- ^ the cut which is to be extended
+    -> cid
+        -- ^ the header onto which the new block is created. It is expected
+        -- that this header is contained in the cut.
+    -> Maybe CutExtension
+getCutExtension c cid = do
+
+    -- In a graph transition we wait for all chains to do the transition
+    -- to the new graph before moving ahead.
+    guard (not $ isGraphTransitionCut && isGraphTransitionPost)
+
+    -- If this is a graph transition step we have to check braiding for the
+    -- old /and/ the new graph.
+    when isGraphTransitionPre (void $ newAdjHashes parentGraph)
+
+    -- Finally, compute the adjacent hashes for the target graph.
+    as <- BlockHashRecord <$> newAdjHashes targetGraph
+
+    return CutExtension
+        { _cutExtensionCut' = c
+        , _cutExtensionParent' = ParentHeader p
+        , _cutExtensionAdjacentHashes' = as
+        }
+  where
+    p = c ^?! ixg (_chainId cid)
+    v = _chainwebVersion c
+    parentHeight = _blockHeight p
+    targetHeight = parentHeight + 1
+    parentGraph = chainGraphAt_ p parentHeight
+    targetGraph = chainGraphAt_ p targetHeight
+    isGraphTransitionPre = isGraphChange c targetHeight
+    isGraphTransitionPost = isGraphChange c parentHeight
+    isGraphTransitionCut = isTransitionCut c
+
+    -- | Try to get all adjacent hashes dependencies for the given graph.
+    --
+    newAdjHashes :: ChainGraph -> Maybe (HM.HashMap ChainId BlockHash)
+    newAdjHashes g =
+        imapM (\xcid _ -> hashForChain xcid)
+        $ HS.toMap
+        $ adjacentChainIds g p
+
+    hashForChain acid
+        -- existing chain
+        | Just b <- c ^? ixg acid = tryAdj b
```

Comment:
```suggestion
        | [something legible] = tryAdj b
```
1. I had to look up `<-`. Not a fan, although that's not what's motivating my complaint.
2. Can we get a utility function for the ubiquitous `c ^? ixg foo`? It makes reading this codebase painful.